File: Codeforces root folder2/Boarding Queue.cpp
// python code hit TLE, so cpp, the idea is obvious, but implementation
// is tedious
void solve() {
    int row, col, n, p; cin >> row >> col >> n >> p;
    int grid[row][col];
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            cin >> grid[i][j];
        }
    }
    unordered_set<int> st;
    for (int r = 0; r < row; r++) {
        for (int c = 0; c < col; c++) {
            int val1 = grid[r][c];
            if (val1 <= p && val1 != 0) {
                int coors[4][2] = {{r+1,c},{r-1,c},{r,c+1},{r,c-1}};
                for (auto coor : coors) {
                    int R = coor[0], C = coor[1];
                    if (
                        R >= 0 && R < row &&
                        C >= 0 && C < col &&
                        grid[R][C] != 0
                    ) {
                        int val2 = grid[R][C];
                        int diff = val2 - val1;
                        int cand = p + diff;
                        if (cand >= 1 and cand <= n) {
                            st.insert(cand);
                        }
                    }
                }
            }
        }
    }
    cout << st.size() << '/' << n-1 << '\n' << endl;
}

int main() {
    int t; cin >> t;
    for (int i = 0; i < t; i++) {
        solve();
    }
}

File: Codeforces root folder 2/Boarding Queue.py
# tle on testcase 59
def solve():
    row, col, n, p = [int(e) for e in input().split()]
    grid = []
    for i in range(row):
        grid.append([int(e) for e in input().split()])

    coor = None
    for r in range(row):
        for c in range(col):
            if grid[r][c] == p:
                coor = (r, c)
                break
        else:
            continue
        break

    hashSet = set()
    while True:
        r, c = coor
        val = grid[r][c]
        nextCoor = None
        for R, C in [(r+1, c), (r, c+1), (r-1, c), (r, c-1)]:
            if (
                R in range(row) and
                C in range(col) and
                grid[R][C] != 0
            ):
                val2 = grid[R][C]
                diff = val2 - val
                if p+diff in range(1, n+1):
                    hashSet.add(p+diff)
                if val2 == val-1:
                    nextCoor = (R, C)
        coor = nextCoor
        if val == 1:
            break
    print(str(len(hashSet))+'/'+str(n-1))
solve()

